### PR TITLE
Replace grants.subject with grants.subject_id and grants.subject_kind

### PR DESCRIPTION
**Pull request by dnephin**
## Summary

Best viewed by individual commit.

Migrates the `grants` table from an encoded `subject` field to `subject_id bigint` and `subject_kind smallint`  fields.

Also update all the places that set `models.Grant.Subject` to use the new `models.Subject` type. This removes the last use of `uid.PolymorphicID`.

## Related Issues

Related to https://github.com/infrahq/infra/issues/3761#issuecomment-1380703072

### CODE REVIEW

File: internal/server/data/grant.go [hunk 1]
```diff
@@ -104,8 +99,9 @@ func GetGrant(tx ReadTxn, opts GetGrantOptions) (*models.Grant, error) {
 	switch {
 	case opts.ByID != 0:
 		query.B("AND id = ?", opts.ByID)
-	case opts.BySubject != "":
-		query.B("AND subject = ?", opts.BySubject)
+	case opts.BySubject.ID != 0:
```
Comment (BruceMacD) [inline internal/server/data/grant.go:102]: Looks like you should be requiring subject kind here?

Comment (dnephin) [reply internal/server/data/grant.go:102]: Good call, I was kind of lazy with these checks. I guess all of these should error if there is an ID set without a kind. The query it produces is still safe in that it matches nothing, but an explicit error will make it easier to debug when we make a mistake.

**Comment by dnephin** [comment]
Tests are passing, and the migration test has an expected value, so I think this is ready for review.

This change required a few steps, so best viewed by individual commit.

File: internal/server/data/migrations.go [hunk 2]
```diff
@@ -1173,3 +1174,70 @@ func addUserPublicKeyUserIDIndex() *migrator.Migration {
...
+				if err != nil {
+					return err
+				}
+			}
+
+			// Step 3 - Remove the old column, add a new index, and NOT NULL
+			_, err = tx.Exec(`
+				ALTER TABLE grants DROP COLUMN IF EXISTS subject;
+				ALTER TABLE grants ALTER COLUMN subject_id SET NOT NULL;
+				ALTER TABLE grants ALTER COLUMN subject_kind SET NOT NULL;
```
Comment (dnephin) [inline internal/server/data/migrations.go:1236]: This final step in the migration will cause problems for the other pods (that have not yet been updated to the new code). We could save this step for a later commit, or we could accept the few errors while the pods restart.

Saving it for a later commit only works for SaaS. We'd have to spread this out over 2 releases for on-prem (or accept a few errors during upgrade).

File: internal/server/data/migrations.go [hunk 3]
```diff
@@ -1173,3 +1174,70 @@ func addUserPublicKeyUserIDIndex() *migrator.Migration {
...
+					kind = 1
+				} else {
+					kind = 2
+				}
+				subjectID, err = uid.Parse([]byte(grant.OriginalSubject[2:]))
+				if err != nil {
+					return err
+				}
+
+				_, err = tx.Exec(`UPDATE grants SET subject_id=?, subject_kind=? WHERE id = ?`,
```
Comment (dnephin) [inline internal/server/data/migrations.go:1225]: Iterating over every grant and doing an individual update is relatively slow. We can see how long this takes in dev. I suspect it'll be fine at our current scale.

**Review by mxyng — APPROVED** [review]

File: internal/server/data/migrations.go (hunk 2 above)
Comment (mxyng) [reply internal/server/data/migrations.go:1236]: A few errors in dying pods is likely fine. It doesn't make sense to delay this switchover to mitigate a few seconds of errors

**Pull request merged**